keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,202 @@
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to understand the impact of these changes, we need to understand the their performance implications. We will perform performance/benchmark tests on this change and backward compatibility test.
+
+ Compared to the time that inter-pod affinity was introduced, performance of the scheduler has been greatly improved(https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
+ ## Design Details
+ ### Content
+  We will abstract the `Node` and `Pod` label selection feature, and finally the `NodeSelector` and `PodSelector` will be based on the `LabelSelectorRequirement` implementation:
+
+ The definition of `NodeSelector` will not change,it is defined as below:
+ ```go
+type NodeSelector struct {
+  NodeSelectorTerms []NodeSelectorTerm
+}
+```
+
+For the definition of `NodeSelectorTerm`, today the api looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []NodeSelectorRequirement
+  MatchFields []NodeSelectorRequirement
+}
+```
+
+We will use `LabelSelectorRequirement` to replace the original `NodeSelectorRequirement` as the base selector of `NodeSelector`. So it looks like this:
+ ```go
+type NodeSelectorTerm struct {
+  MatchExpressions []LabelSelectorRequirement
+  MatchFields []LabelSelectorRequirement
+}
+```
+
+`PodSelector` is our new selector. A pod selector represents the union of the results of one or more label queries over a set of pods. The of `PodSelector` is defined as below:
+ ```go
+type PodSelector struct {
+  MatchLabels map[string]string
+  MatchExpressions []LabelSelectorRequirement
+}
+```
+
+`LabelSelectorRequirement` is an abstract selector implementation of `NodeSelector` and `PodSelector`. And the `LabelSelectorRequirement` is defined as below:
+ ```go
+type LabelSelectorRequirement struct {
+  Key string
+  Operator LabelSelectorOperator
+  Values []string
+}
+
+type LabelSelectorOperator string
+
+const (
+  LabelSelectorOpIn           LabelSelectorOperator = "In"
+  LabelSelectorOpNotIn        LabelSelectorOperator = "NotIn"
+  LabelSelectorOpExists       LabelSelectorOperator = "Exists"
+  LabelSelectorOpDoesNotExist LabelSelectorOperator = "DoesNotExist"
+  LabelSelectorOpGt           LabelSelectorOperator = "Gt"
+  LabelSelectorOpLt           LabelSelectorOperator = "Lt"
+)
+```
+
+ ### Test Plan
+
+The feature has correctness, integration, performance, and e2e tests. These tests are run regularly as a part of Kubernetes presubmit and CI/CD pipeline.
+
+#### Correctness Tests
+Here is a list of unit tests for various modules of the feature:
+
+- `NodeSelector` related tests
+- `PodSelector` related tests
+- `Gt` and `Lt` functional tests
+- `LabelSelectorRequirement` related tests
```

Comment:
we also need tests ensuring that the new Gt/Lt operators are not permitted in "normal" selector fields (e.g. pod.spec.nodeSelector, etc)

Reply:
Yes, this is for sure, pod.spec.nodeSelector(https://github.com/kubernetes/kubernetes/blob/22c67bb113dee2ce0858075cd5af5d110c255985/pkg/apis/core/types.go#L2599) is a standalone map structure.

Reply:
Backwards compatibility - pods made with the new types should still be updatable if apiserver version is rolled back.
Forwards compatibility - all pods created today are wire-compatible (both **proto and json**) with the new api.

Reply:
I agree with your comment. But what needs to be clarified is that because we have added `Gt` and `Lt` operators, this does not exist in the previous API.